src/data-workspace/use-data-value-mutation.js

```diff
@@ -0,0 +1,88 @@
+import { useDataEngine } from '@dhis2/app-runtime'
+import { useQueryClient, useMutation } from 'react-query'
+import { useContextSelection } from '../context-selection/index.js'
+import { dataValueQuery, useAttributeOptionCombo } from './data-workspace.js'
+
+const DATA_VALUE_MUTATION = {
+    resource: 'dataValues',
+    type: 'create',
+    params: ({ ...params }) => ({ ...params }),
+}
+
+export const useDataValueMutation = () => {
+    const queryClient = useQueryClient()
+    const [{ dataSetId, orgUnitId, periodId }] = useContextSelection()
+    const attributeOptionComboId = useAttributeOptionCombo()
+    const engine = useDataEngine()
+
+    const dataValueQueryKey = [
+        dataValueQuery,
+        {
+            dataSetId,
+            periodId,
+            orgUnitId,
+            attributeOptionComboId,
+        },
+    ]
+    const mutationFn = (variables) =>
+        engine.mutate(DATA_VALUE_MUTATION, { variables })
+
+    return useMutation(mutationFn, {
+        // Optimistic update of the react-query cache
+        onMutate: async (newDataValue) => {
+            // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+            await queryClient.cancelQueries(dataValueQueryKey)
+
+            // Snapshot the previous value
+            const previousDataValues =
+                queryClient.getQueryData(dataValueQueryKey)
+
+            // Optimistically update to the new value
+            queryClient.setQueryData(dataValueQueryKey, (oldDataValues) => {
+                const newDataValues = { ...oldDataValues }
+
+                newDataValues.dataValues.dataValues =
```

Comment:
I suppose it doesn't really matter, but this will also update the `oldDataValues.dataValues` property as the referencing the same object

Reply:
Yeah I was wondering the same. I'll double check it, because it could affect the snapshot.

Reply:
Good thing you commented on this, I'd glossed over it, but this would have caused a bug. The oldDataValues and previousDataValues are both references to the same object. So mutating oldDataValues would have mutated the snapshot and thus broken the rollback. Addressed it with cfdc56c